relay-system/src/service.rs

```diff
@@ -44,45 +116,315 @@ impl fmt::Display for SendError {
+    }
+}
+
+impl MessageResponse for NoResponse {
+    type Sender = ();
+    type Output = ();
+
+    fn channel() -> (Self::Sender, Self::Output) {
+        ((), ())
+    }
+}
+
+/// Declares a message as part of an [`Interface`].
+///
+/// Messages have an associated `Response` type that determines the return value of sending the
+/// message. Within an interface, the responder can vary for each message. There are two provided
+/// responders.
+///
+/// # No Response
+///
+/// [`NoResponse`] is used for fire-and-forget messages that do not return any values. These
+/// messages do not spawn futures and cannot be awaited. It is neither possible to verify whether
+/// the message was delivered to the service.
+///
+/// When implementing `FromMessage` for such messages, the second argument can be ignored by
+/// convention:
+///
+/// ```
+/// use relay_system::{FromMessage, Interface, NoResponse};
+///
+/// struct MyMessage;
+///
+/// enum MyInterface {
+///     MyMessage(MyMessage),
+///     // ...
+/// }
+///
+/// impl Interface for MyInterface {}
+///
+/// impl FromMessage<MyMessage> for MyInterface {
+///     type Response = NoResponse;
+///
+///     fn from_message(message: MyMessage, _: ()) -> Self {
+///         Self::MyMessage(message)
+///     }
+/// }
+/// ```
+///
+/// # Asynchronous Responses
+///
+/// [`AsyncResponse`] is used for messages that resolve to some future value. This value is sent
+/// back by the service through a [`Sender`], which must be added into the interface:
+///
+/// ```
+/// use relay_system::{AsyncResponse, FromMessage, Interface, Sender};
+///
+/// struct MyMessage;
+///
+/// enum MyInterface {
+///     MyMessage(MyMessage, Sender<bool>),
+///     // ...
+/// }
+///
+/// impl Interface for MyInterface {}
+///
+/// impl FromMessage<MyMessage> for MyInterface {
+///     type Response = AsyncResponse<bool>;
+///
+///     fn from_message(message: MyMessage, sender: Sender<bool>) -> Self {
+///         Self::MyMessage(message, sender)
+///     }
+/// }
+/// ```
+///
+/// See [`Interface`] for more examples on how to build interfaces using this trait.
+pub trait FromMessage<M>: Interface {
+    /// The behavior declaring the return value when sending this message.
+    type Response: MessageResponse;
+
+    /// Converts the message into the service interface.
+    fn from_message(message: M, sender: <Self::Response as MessageResponse>::Sender) -> Self;
+}
+
+/// The address of a [`Service`].
 ///
-/// The address of a [`Service`] allows you to [send](Addr::send) messages to the service as
-/// long as the service is running.  It can be freely cloned.
-#[derive(Debug)]
-pub struct Addr<S: Service> {
-    /// The transmitter of the channel used to communicate with the Service
-    pub tx: mpsc::UnboundedSender<S::Messages>,
+/// The address of a [`Service`] allows you to [send](Self::send) messages to the service as
+/// long as the service is running. It can be freely cloned.
+pub struct Addr<I: Interface> {
```

Comment:
Out of curiosity, would it be possible to eliminate the `Interface` trait by writing something like

```rust
pub struct Addr<M, I: FromMessage<M>> {
```

Reply:
Yes, that's possible.  It kind of was like that before this refactor as well.  But we ended up going with explicitly marking this with a trait mostly to give it a name and clearly designate it's purpose.